crypto/bio/bio_lib.c

```diff
@@ -12,26 +12,54 @@
 #include <openssl/crypto.h>
 #include "bio_lcl.h"
 #include "internal/cryptlib.h"
-#include <openssl/stack.h>
+
+/* internal BIO init wrapper */
+static int bio_init(BIO *bio, const BIO_METHOD *method);
+static void bio_clear(BIO *bio);
 
 BIO *BIO_new(const BIO_METHOD *method)
 {
-    BIO *ret = OPENSSL_malloc(sizeof(*ret));
+    BIO *bio = OPENSSL_zalloc(sizeof(*bio));
 
-    if (ret == NULL) {
+    if (bio == NULL) {
         BIOerr(BIO_F_BIO_NEW, ERR_R_MALLOC_FAILURE);
         return (NULL);
     }
-    if (!BIO_set(ret, method)) {
-        OPENSSL_free(ret);
-        ret = NULL;
+
+    if (!bio_init(bio, method)) {
+        OPENSSL_free(bio);
+        bio = NULL;
     }
-    return (ret);
+    return bio;
 }
 
+/*
+ * Was useful to setup/init any stack-allocated BIO object.
+ * Its use is now deprecated as : 
+ *  - BIO type is now opaque, so no more stack-allocable.
+ *  - new locking API uses a now-dynamic allocated member to lock 
+ *  each BIO object individually.
+ *
+ * Warning: Possible leaks of [lock| and |ex_data| return fields...
+ */
+# if OPENSSL_API_COMPAT < 0x10100000L
 int BIO_set(BIO *bio, const BIO_METHOD *method)
 {
-    bio->method = method;
+    /* These two existing fields will leak else */
+    OPENSSL_assert(bio->lock == NULL);
+    OPENSSL_assert(bio->ex_data.sk == NULL);
```

Comment:
Actually....thinking about this some more, isn't it the case that as far as user code is concerned we can *never* get a situation where these two are NULL? They get created during BIO_new and freed during BIO_free, and BIO's are not stack allocatable any more. That means these asserts will *always* be triggered, and therefore the application will crash...which is a "bad thing". Would it be better to remove BIO_set altogether since there is never a situation where it can be sensibly used?